src/core/coap/coap.cpp

```diff
@@ -620,35 +620,39 @@ void CoapBase::ProcessReceivedRequest(Message &aMessage, const Ip6::MessageInfo
     }
 }
 
-CoapMetadata::CoapMetadata(bool                    aConfirmable,
-                           const Ip6::MessageInfo &aMessageInfo,
-                           otCoapResponseHandler   aHandler,
-                           void *                  aContext)
+void CoapBase::CoapMetadata::Init(bool                    aConfirmable,
+                                  const Ip6::MessageInfo &aMessageInfo,
+                                  otCoapResponseHandler   aHandler,
+                                  void *                  aContext)
 {
+    uint32_t maxRandom = Time::SecToMsec(kAckTimeout) * kAckRandomFactorNumerator / kAckRandomFactorDenominator -
+                         Time::SecToMsec(kAckTimeout) + 1;
+
     mSourceAddress         = aMessageInfo.GetSockAddr();
-    mDestinationPort       = aMessageInfo.GetPeerPort();
     mDestinationAddress    = aMessageInfo.GetPeerAddr();
+    mDestinationPort       = aMessageInfo.GetPeerPort();
     mResponseHandler       = aHandler;
     mResponseContext       = aContext;
+    mRetransmissionTimeout = Time::SecToMsec(kAckTimeout) + Random::NonCrypto::GetUint32InRange(0, maxRandom);
     mRetransmissionCount   = 0;
-    mRetransmissionTimeout = Time::SecToMsec(kAckTimeout);
-    mRetransmissionTimeout += Random::NonCrypto::GetUint32InRange(
-        0, Time::SecToMsec(kAckTimeout) * kAckRandomFactorNumerator / kAckRandomFactorDenominator -
-               Time::SecToMsec(kAckTimeout) + 1);
+    mAcknowledged          = false;
+    mConfirmable           = aConfirmable;
+    mNextTimerShot = TimerMilli::GetNow() + (aConfirmable ? mRetransmissionTimeout : Time::SecToMsec(kMaxTransmitWait));
+}
 
-    if (aConfirmable)
-    {
-        // Set next retransmission timeout.
-        mNextTimerShot = TimerMilli::GetNow() + mRetransmissionTimeout;
-    }
-    else
-    {
-        // Set overall response timeout.
-        mNextTimerShot = TimerMilli::GetNow() + Time::SecToMsec(kMaxTransmitWait);
-    }
+void CoapBase::CoapMetadata::ReadFrom(const Message &aMessage)
+{
+    uint16_t length;
+
+    length = aMessage.Read(aMessage.GetLength() - sizeof(*this), sizeof(*this), this);
+    assert(length == sizeof(*this));
 
-    mAcknowledged = false;
-    mConfirmable  = aConfirmable;
+    OT_UNUSED_VARIABLE(length);
```

Comment:
nit: prefer at top of method body in own block

Reply:
I actually changed it fully to remove the `OT_UNUSED_VARIABLE()` (by using the `length`).

I was not sure about the value of `assert(length == sizeof(*this))` in the original code, since it is basically verifying the behavior of `Message::Read()`.
```
    {
        uint16_t length = aMessage.Read(aMessage.GetLength() - sizeof(*this), sizeof(*this), this);
        assert(length == sizeof(*this));
        OT_UNUSED_VARIABLE(length);
    }
```

I changed it to this which instead `assert`s that `aMessage` is long enough to contain the metadata before reading it.
```
void CoapBase::CoapMetadata::ReadFrom(const Message &aMessage)
{
    uint16_t length = aMessage.GetLength();

    assert(length >= sizeof(*this));
    aMessage.Read(length - sizeof(*this), sizeof(*this), this);
}
```